index.html

```diff
@@ -140,67 +166,250 @@ <h2>
+          Promise&lt;void&gt; clearAppBadge();
         };
+
+        Navigator includes NavigatorBadge;
+        WorkerNavigator includes NavigatorBadge;
       </pre>
       <p>
-        The <dfn>Badge</dfn> namespace provides functions for setting and
-        clearing the application badge indicator.
+        User agents that never display document badges SHOULD NOT expose the
+        {{Navigator/setClientBadge}} and {{Navigator/clearClientBadge}}
+        methods. Similarly, user agents that never display application badges
+        SHOULD NOT expose the {{NavigatorBadge/setAppBadge}} and
+        {{NavigatorBadge/clearAppBadge}} methods.
+      </p>
+      <p class="note">
+        This is important as a feature-detection mechanism to allow sites to
+        determine whether setting a badge will have any effect, based on the
+        presence of the methods. In particular, sites can use the absence of
+        {{Navigator/setClientBadge}} as a condition for displaying a fallback
+        badge in the page's <a data-cite=
+        "html#the-title-attribute"><code>title</code></a> or <a data-cite=
+        "html#rel-icon"><code>icon</code></a>.
+      </p>
+      <p class="note">
+        These methods return a {{Promise}}, even though there are no
+        asynchronous failure conditions (the only failure conditions are WebIDL
+        type errors). This is to allow us to add asynchronous failure
+        conditions in the future without altering how type errors are returned.
       </p>
       <section>
         <h3>
-          <dfn>set()</dfn> function
+          <dfn data-dfn-for="Navigator">setClientBadge()</dfn> method
         </h3>
         <p>
-          When the <code>set()</code> function is called with argument
+          When the <code>setClientBadge()</code> method is called with argument
           <var>contents</var>, run the following steps:
         </p>
         <ol>
-          <li>If <var>contents</var> is omitted, set the application's badge to
-          <a>flag</a>.
+          <li>Let <var>document</var> be the <a data-cite=
+          "html#current-settings-object">current settings object</a>'s
+          <a data-cite="html#responsible-document">responsible document</a>.
+          </li>
+          <li>If <var>contents</var> is omitted, set the badge associated with
+          <var>document</var> to <a>flag</a>.
           </li>
-          <li>If <var>contents</var> is 0, set the application's badge to
-          <a>nothing</a>.
+          <li>Else, if <var>contents</var> is 0, set the badge associated with
+          <var>document</var> to <a>nothing</a>.
           </li>
-          <li>Else, set the application's badge to <var>contents</var>.
+          <li>Else, set the badge associated with <var>document</var> to <var>
+            contents</var>.
+          </li>
+          <li>Return a <a data-cite="webidl#a-promise-resolved-with">resolved
```

Comment:
Hurray for promises eh? :cry:

Reply:
This should be, in parallel, attempt to set the badge. If setting the badge succeeds, then resolve the promise. Otherwise, if failure, reject promise with an appropriate DOMException. 

That's how I plan to implement this in Gecko... I'll try to do the actual IPC, then resolve only when I get back an affirmative from that the OS got the message (and possibly succeeded). 

There are a bunch of reasons why this could fail... we just haven't teased out what they are yet! 

One example, if the OS detects you are trying to animate the badge by sending too many updates, it may block you with a `SecurityError` DOMException.

Reply:
OK so I've rewritten this part of the spec. For now, it's "fire and forget", so we return the resolved promise immediately, and set the badge in parallel. (As opposed to resolving/rejecting the promise later based on the success in the OS layer.) Please discuss on #61.